perfkitbenchmarker/benchmarks/fio_benchmark.py

```diff
@@ -54,6 +113,122 @@
                   'num_machines': 1}
 
 
+GLOBALS_TEMPLATE = """
+[global]
+ioengine=libaio
+invalidate=1
+direct=1
+runtime=10m
+time_based
+filename={{filename}}
+do_verify=0
+verify_fatal=0
+randrepeat=0
+"""
+
+JOB_TEMPLATE = """
+{% for iodepth in iodepths %}
+[{{name}}-io-depth-{{iodepth}}]
+stonewall
+rw={{rwkind}}
+blocksize={{blocksize}}
+iodepth={{iodepth}}
+{% endfor %}
+"""
+
+
+def GetIODepths(io_depths):
+  """Parse the io_depths parameter.
+
+  Args:
+    io_depths: a string in the format of the --io_depths flag.
+
+  Returns:
+    An iterable of integers.
+
+  Raises:
+    ValueError if io_depths doesn't follow a format it recognizes.
+  """
+
+  match = IODEPTHS_REGEXP.match(io_depths)
+
+  if match.group(2) is None:
+    return [int(match.group(1))]
+  else:
+    return range(int(match.group(1)), int(match.group(3)) + 1)
+
+
+def GenerateJobFileString(disk, against_device,
+                          scenarios, io_depths):
+  """Write a fio job file.
+
+  Args:
+    disk: the disk.BaseDisk object we're benchmarking with.
+    against_device: bool. True if we're using a raw disk.
+    scenarios: iterable of dicts, taken from SCENARIOS.
+    io_depths: iterable. The IO queue depths to test.
+
+  Returns:
+    The contents of a fio job file, as a string.
+  """
+
+  if against_device:
+    filename = disk.GetDevicePath()
+  else:
+    filename = posixpath.join(disk.mount_point, DEFAULT_TEMP_FILE_NAME)
+
+  globals_template = jinja2.Template(GLOBALS_TEMPLATE,
+                                     undefined=jinja2.StrictUndefined)
+  job_template = jinja2.Template(JOB_TEMPLATE,
+                                 undefined=jinja2.StrictUndefined)
+
+  file_string = str(globals_template.render(filename=filename))
+  for scenario in scenarios:
+    file_string = file_string + str(job_template.render(
+        name=scenario['name'],
+        rwkind=scenario['rwkind'],
+        blocksize=scenario['blocksize'],
+        iodepths=io_depths))
+
+  return file_string
+
+
+def JobFileString(fio_jobfile, disk, against_device,
+                  scenario_strings, io_depths):
+  """Get the contents of our job file.
+
+  Args:
+    fio_jobfile: string or False. The path to the user's jobfile, if provided.
+    disk: the disk.BaseDisk object we're benchmarking with.
+    against_device: bool. True if we're using a raw disk.
+    scenario_strings: iterable of strings. The workload scenarios to generate.
+    io_depths: iterable. The IO queue depths to test.
+
+    vm: the virtual_machine.BaseVirtualMachine that we will run on.
+
+  Returns:
+    A string containing the user's job file.
+  """
+
+  # The default behavior is to run with the standard fio job file.
+  if scenario_strings is None and not fio_jobfile:
+    fio_jobfile = data.ResourcePath('fio.job')
+
+  if fio_jobfile:
+    with open(fio_jobfile, 'r') as jobfile:
+      return jobfile.read()
+  else:
```

Comment:
The else block can be executed if the user passes --generate_scenarios and does not pass --fio_jobfile. In that case, scenario_strings will either be [] or a list of strings and line 214 will not reset fio_jobfile, so it is possible to hit the else case. I will make the code more clear.